OpenRA.Mods.Common/Traits/BotModules/BridgeRepairBotModule.cs

```diff
@@ -0,0 +1,156 @@
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Pathfinder;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages AI legacy bridge repair logic.")]
+	public class BridgeRepairBotModuleInfo : ConditionalTraitInfo
+	{
+		[ActorReference]
+		[Desc("Actor types that can repair bridges (via `RepairsBridges`).",
+			"Leave this empty to disable the trait.")]
+		public readonly HashSet<string> RepairActorTypes = new HashSet<string>();
+
+		[Desc("Avoid enemy actors nearby when searching for bridges in need of repair.",
+			"Should be somewhere near the max weapon range.")]
+		public readonly WDist EnemyAvoidanceRadius = WDist.FromCells(8);
+
+		[Desc("Minimum delay (in ticks) between trying to scan for repair targets.")]
+		public readonly int MinimumWaitDelay = 300;
+
+		public override object Create(ActorInitializer init) { return new BridgeRepairBotModule(init.Self, this); }
+	}
+
+	public class BridgeRepairBotModule : ConditionalTrait<BridgeRepairBotModuleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		readonly Predicate<Actor> unitCannotBeOrderedOrIsIdle;
+
+		// Units that the bot already knows about and has given a repair order. Any unit not on this list needs to be given a new order.
+		readonly List<Actor> activeRepairers = new List<Actor>();
+
+		int waitDelayTicks;
+
+		IPathFinder pathfinder;
+		DomainIndex domainIndex;
+
+		IBotRequestUnitProduction[] requestUnitProduction;
+
+		public BridgeRepairBotModule(Actor self, BridgeRepairBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+
+			if (world.Type == WorldType.Editor)
+				return;
+
+			unitCannotBeOrderedOrIsIdle = a => a.Owner != player || a.IsDead || !a.IsInWorld || a.IsIdle;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			waitDelayTicks = world.LocalRandom.Next(Info.MinimumWaitDelay);
+
+			pathfinder = world.WorldActor.Trait<IPathFinder>();
+			domainIndex = world.WorldActor.Trait<DomainIndex>();
+
+			requestUnitProduction = player.PlayerActor.TraitsImplementing<IBotRequestUnitProduction>().ToArray();
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			if (--waitDelayTicks <= 0)
+			{
+				waitDelayTicks = Info.MinimumWaitDelay;
+				QueueRepairOrders(bot);
+			}
+		}
+
+		void QueueRepairOrders(IBot bot)
+		{
+			if (!Info.RepairActorTypes.Any() || player.WinState != WinState.Undefined)
+				return;
+
+			activeRepairers.RemoveAll(unitCannotBeOrderedOrIsIdle);
+
+			var newUnits = world.ActorsHavingTrait<IPositionable>()
+				.Where(a => a.Owner == player && !activeRepairers.Contains(a));
+
+			var repairers = newUnits
```

Comment:
These are heavy queries that will almost never be needed (many maps don't have bridges, and those that do are almost never broken) - move it after the `if (!targetOptions.Any())` check.